statistics/handle/handle.go

```diff
@@ -807,3 +850,103 @@ func (h *Handle) releaseStatsReader(reader *statsReader) error {
 	h.mu.Unlock()
 	return err
 }
+
+const (
+	// StatsStatusInited is the status for extended stats which are just registered but have not been analyzed yet.
+	StatsStatusInited uint8 = iota
+	// StatsStatusAnalyzed is the status for extended stats which have been collected in analyze.
+	StatsStatusAnalyzed
+	// StatsStatusDeleted is the status for extended stats which were dropped. These "deleted" records would be removed from storage by GCStats().
+	StatsStatusDeleted
+)
+
+// InsertExtendedStats inserts a record into mysql.stats_extended and update version in mysql.stats_meta.
+func (h *Handle) InsertExtendedStats(statsName, db string, colIDs []int64, tp int, tableID int64, ifNotExists bool) (err error) {
+	bytes, err := json.Marshal(colIDs)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	strColIDs := string(bytes)
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	ctx := context.TODO()
+	exec := h.mu.ctx.(sqlexec.SQLExecutor)
+	_, err = exec.Execute(ctx, "begin pessimistic")
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer func() {
+		err = finishTransaction(ctx, exec, err)
+	}()
+	txn, err := h.mu.ctx.Txn(true)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	version := txn.StartTS()
+	sql := fmt.Sprintf("INSERT INTO mysql.stats_extended(stats_name, db, type, table_id, column_ids, version, status) VALUES ('%s', '%s', %d, %d, '%s', %d, %d)", statsName, db, tp, tableID, strColIDs, version, StatsStatusInited)
+	_, err = exec.Execute(ctx, sql)
+	// Key exists, but `if not exists` is specified, so we ignore this error.
+	if kv.ErrKeyExists.Equal(err) && ifNotExists {
+		err = nil
+	}
+	return
+}
+
+// MarkExtendedStatsDeleted update the status of mysql.stats_extended to be `deleted` and the version of mysql.stats_meta.
+func (h *Handle) MarkExtendedStatsDeleted(statsName, db string, tableID int64) (err error) {
+	if tableID < 0 {
+		sql := fmt.Sprintf("SELECT table_id FROM mysql.stats_extended WHERE stats_name = '%s' and db = '%s'", statsName, db)
+		rows, _, err := h.restrictedExec.ExecRestrictedSQL(sql)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if len(rows) == 0 {
+			return nil
+		}
+		tableID = rows[0].GetInt64(0)
+	}
```

Comment:
If two stats name of different tables collide, does this logic cause something wrong?
For example:
```
CREATE STATISTICS statsABC (cardinality) ON t1(a, b, c);
CREATE STATISTICS statsABC (cardinality) ON t2(a, b, c);
DROP STATISTICS statsABC;
```

Reply:
Currently, `(db, statsName)` is a unique constraint defined in the table schema.